### PR TITLE
Added short code for route cipher

### PR DESCRIPTION
**Pull request by Saptashrungi**
#### References to other Issues or PRs

https://github.com/sympy/sympy/issues/16543

#### Release Notes

* crypto
  * Added Route Cipher

### CODE REVIEW

**Comment by namannimmo10** [comment]
You should try it this way:
```
<!-- BEGIN RELEASE NOTES -->
* crypto
    *  Added a short code for route cipher. On further modification it can be used for key with any of 
       following parameters: clockwise/anticlockwise, starting point and inwards/outwards direction.
<!-- END RELEASE NOTES -->
```

File: sympy/crypto/route_cipher.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    elif(c==4):
+        return(l4+l1+l2+l3)  #4 when spiral starts from bottom left corner
+o=[]
+for i in range(2): #2 is no of columns/2
+    o+=([j for j in create(i,i,3,4,4)])
+print([i for i in o if i is not None])    
+    
+#this way all cases can be considered, the key will tell from where to start
+#for anti-clockwise i will use reverse
+#also clockwise inwards will be reverse of anticlockwise outwards spiral, so inwards and outwards both can be considered
```
Comment (czgdp1807) [inline sympy/crypto/route_cipher.py]: Add a new line at the end of the file.

File: sympy/crypto/route_cipher.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            mat[i][l]=None
+    if(c==1):
+        return(l1+l2+l3+l4)  #1 when clockwise spiral starts from top-left corner
+    elif(c==2):
+        return(l2+l3+l4+l1)  #2 when spiral starts from top-right corner
+    elif(c==3):
+        return(l3+l4+l2+l1)  #3 when spiral starts from bottom right corner
+    elif(c==4):
+        return(l4+l1+l2+l3)  #4 when spiral starts from bottom left corner
+o=[]
```
Comment (czgdp1807) [inline sympy/crypto/route_cipher.py]: Space around `=`

File: sympy/crypto/route_cipher.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+for i in range(2): #2 is no of columns/2
+    o+=([j for j in create(i,i,3,4,4)])
+print([i for i in o if i is not None])
```
Comment (czgdp1807) [inline sympy/crypto/route_cipher.py]: No `print()` statements

File: sympy/crypto/route_cipher.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            l4.append(mat[i][l])   #creates a list from 7 to 4
+            mat[i][l]=None
+    if(c==1):
+        return(l1+l2+l3+l4)  #1 when clockwise spiral starts from top-left corner
+    elif(c==2):
+        return(l2+l3+l4+l1)  #2 when spiral starts from top-right corner
+    elif(c==3):
+        return(l3+l4+l2+l1)  #3 when spiral starts from bottom right corner
+    elif(c==4):
+        return(l4+l1+l2+l3)  #4 when spiral starts from bottom left corner
```
Comment (czgdp1807) [inline sympy/crypto/route_cipher.py]: No parenthesis with if-else-elif.

**Comment by czgdp1807** [comment]
Please DO NOT clear the PULL REQUEST TEMPLATE.

**Comment by czgdp1807** [comment]
A better way to add route cipher would be to add the code in https://github.com/sympy/sympy/blob/master/sympy/crypto/crypto.py

From the current pattern in `sympy.crypto.crypto`, there is a combination of encryption and decryption functions, I suggest to use the following functions for route cipher,

1. `encipher_route(message, cols, route)`
2. `decipher_route(message, cols, route)`

Here `route` can be used to denote the way we traverse the grid of text and `cols` denotes the number of columns in the  grid.
Refer, https://crypto.interactive-maths.com/route-cipher.html

**Comment by smichr** [comment]
> A better way to add route cipher would be to add the code in

I interpreted the current work as a proof of concept looking for feedback. But I agree with the direction you are suggesting.

File: sympy/crypto/route_cipher.py (hunk 4 above)
Comment (Saptashrungi) [reply sympy/crypto/route_cipher.py]: Okay I will do the modifications in the final version.

File: sympy/crypto/route_cipher.py (hunk 3 above)
Comment (Saptashrungi) [reply sympy/crypto/route_cipher.py]: Means?

**Comment by smichr** [comment]
A possibility for "route" designations would be letters (U, D, L, R) and a starting position. The letters would stand for Up/Down columns, Left/Right rows (and UD would mean alternate between up and down while UR or DL would be a clockwise spiral and DR or UL would we be a counter-clockwise spiral. The convention could be that the grid would be adjusted so the initial direction would traverse the entire width/height of the rectangle in which the letters are written. So if there are 20 letters, the columns are 4 and the directive is `U2` the initial grid,

```
0  1  2  3 
4  5  6  7 
8  9  10 11
12 13 14 15
16 17 18 19
```

would be logically changed to 

```
  6  7  4  5
 10 11  8  9
 14 15 12 13
 18 19 16 17
  2  3  0  1
```
and the letters would be read in order 2,18,14,10,6,3,19,15,11,7,0,...

For the same grid, `RD2` would be logically read from
```
  2  3  0  1
  6  7  4  5
 10 11  8  9
 14 15 12 13
 18 19 16 17
```
as 2,3,0,1,5,9,13,17,16,19,18,14,10,6,7,...

**Comment by czgdp1807** [comment]
@Saptashrungi Would you still like to work on this?

**Comment by Saptashrungi** [comment]
Yes, I am back on it now. The encipher function will be done soon.

**Comment by czgdp1807** [comment]
Closing this in favour of https://github.com/sympy/sympy/pull/18644

**Pull request closed without merging**